### PR TITLE
Lockfile base64 checksum support. Use = as checksum algo separator.

### PR DESCRIPTION
**Pull request by martinemde**
## What was the end-user or developer problem that led to this PR?

Follow-up to #6374 that makes this version of Bundler a bit more future proof. I was influenced by the `Content Digest` http header which uses = signs and base64 digests. This just covers us if we choose to change it since the base64 stuff was already here and the `=` is easy now before this is released.

## What is your fix for the problem, implemented in this PR?

Handle base64 checksums and use the `=` character to separate the algo and digest.

Some checksum algorithm names are written with a hypen in them, so using `=` is less likely to cause future problems if we adopt, for example, "SHA3-512".

## Make sure the following tasks are checked

- [X] Describe the problem / feature
- [X] Write [tests](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#tests) for features and bug fixes
- [X] Write code to solve the problem
- [X] Make sure you follow the [current code style](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#code-formatting) and [write meaningful commit messages without tags](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#commit-messages)

### CODE REVIEW

**Review by indirect — APPROVED** [review]

File: bundler/lib/bundler/checksum.rb [hunk 1]
```diff
@@ -15,20 +15,24 @@ def from_gem(io, pathname, algo = DEFAULT_ALGORITHM)
...
-        Checksum.new(algo, digest, Source.new(:lock, lockfile_location))
+        algo, digest = lock_checksum.strip.split(ALGO_SEPARATOR, 2)
+        Checksum.new(algo, to_hexdigest(digest, algo), Source.new(:lock, lockfile_location))
+      end
+
+      def to_hexdigest(digest, algo = DEFAULT_ALGORITHM)
+        return digest unless algo == DEFAULT_ALGORITHM
+        return digest if digest.match?(/\A[0-9a-f]{64}\z/i)
+        if digest.match?(%r{\A[-0-9a-z_+/]{43}={0,2}\z}i)
+          digest.tr("-_", "+/") # fix urlsafe base64
```
Comment (martinemde) [inline bundler/lib/bundler/checksum.rb:31]:
```suggestion
          digest = digest.tr("-_", "+/") # fix urlsafe base64
```

**Review by martinemde — COMMENTED** [review]
Oops

**Comment by martinemde** [comment]
Added a test enforcing that the compact index may return base64 so that at least one test says that we are preserving base64 forwards compatibility.

**Comment by martinemde** [comment]
As an aside here, I found myself wishing that something in the compact index response said what algorithm it was using for the checksums. If it was a header, for example, then we could almost transparently handle any algorithm from this point onwards. I would just need to make sure the parser knew to convert all base64 hashes to hex, or vice versa, before comparing them (or just compare them as binary strings preserving whatever format they are in).

At this point, if we ever wanted to support another algorithm we could just tack on another attribute at the end of the line on compact index in the format of `algo=digest` and it would compare them string to string without any processing for base64. Interesting to think about, but probably far off. SHA256 is quite adequate for what we do.

**Pull request merged**